origin-notifications/app.js

```diff
@@ -0,0 +1,96 @@
+require('dotenv').config()
+
+const express = require('express')
+const bodyParser = require('body-parser')
+const webpush = require('web-push')
+const Sequelize = require('sequelize')
+
+const app = express()
+const port = 3456
+
+const PushSubscription = require('./src/models').PushSubscription
+const emailAddress = process.env.VAPID_EMAIL_ADDRESS
+const privateKey = process.env.VAPID_PRIVATE_KEY
+const publicKey = process.env.VAPID_PUBLIC_KEY
+
+webpush.setVapidDetails(
+  `mailto:${emailAddress}`,
+  publicKey,
+  privateKey
+)
+
+// should be tightened up for security
+app.use((req, res, next) => {
+  res.header("Access-Control-Allow-Origin", "*")
+  res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept")
+
+  next()
+})
+
+app.use(bodyParser.json())
+
+// currently showing all subscriptions in an unauthenticated index view - maybe not a good idea in production?
+app.get('/', async (req, res) => {
+  const subs = await PushSubscription.findAll()
+
+  res.send(`<h1>${subs.length} Push Subscriptions</h1><ul>${subs.map(s => `<li><pre style="white-space: pre-wrap;word-break: break-all">${JSON.stringify(s)}</pre></li>`)}</ul>`)
+})
+
+app.post('/', async(req, res) => {
```

Comment:
We add authentication to this endpoint so that only the account owner can insert data.
Perhaps the message could be signed using the account pk and then the back-end verifies the signature ? Though that would mean a meta-mask popup which is always an annoying user experience... Another possibility would be to create authenticated sessions at the DApp layer so that the user does not have to sign messages every single time she takes an action that requires authentication with the back-end. That could be a good framework to have in place for future ? Let's discuss...

Reply:
Now rate limited, as we agreed that anyone should be able to subscribe to any events ✅